ibtk/src/utilities/SAMRAIDataCache.cpp

```diff
@@ -0,0 +1,297 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibtk/SAMRAIDataCache.h"
+#include "CellDataFactory.h"
+#include "CellVariable.h"
+#include "EdgeDataFactory.h"
+#include "EdgeVariable.h"
+#include "FaceDataFactory.h"
+#include "FaceVariable.h"
+#include "NodeDataFactory.h"
+#include "NodeVariable.h"
+#include "OuteredgeDataFactory.h"
+#include "OuteredgeVariable.h"
+#include "OuterfaceDataFactory.h"
+#include "OuterfaceVariable.h"
+#include "OuternodeDataFactory.h"
+#include "OuternodeVariable.h"
+#include "OutersideDataFactory.h"
+#include "OutersideVariable.h"
+#include "SideDataFactory.h"
+#include "SideVariable.h"
+#include "VariableDatabase.h"
+#include "ibtk/ibtk_utilities.h"
+#include "ibtk/namespaces.h" // IWYU pragma: keep
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBTK
+{
+/////////////////////////////// STATIC ///////////////////////////////////////
+
+namespace
+{
+template <typename U, typename T>
+inline bool
+can_convert_to(T* t)
+{
+    return dynamic_cast<U*>(t) != nullptr;
+}
+
+template <typename U, typename T>
+inline bool
+can_convert_to(const SAMRAI::tbox::Pointer<T>& t)
+{
+    return dynamic_cast<U*>(t.getPointer()) != nullptr;
+}
+
+template <typename T>
+inline bool
+get_data_characteristics(const int idx, int& depth, int& ghost_width)
+{
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    Pointer<Variable<NDIM> > var;
+    var_db->mapIndexToVariable(idx, var);
+    auto patch_descriptor = var_db->getPatchDescriptor();
+
+    // Determine the data depth and ghost width from the patch data factory.
+    //
+    // TODO: Make this more generic and extensible.
+    bool convertable = false;
+    if (can_convert_to<CellVariable<NDIM, T> >(var))
+    {
+        Pointer<CellDataFactory<NDIM, T> > pdat_fac = patch_descriptor->getPatchDataFactory(idx);
+        TBOX_ASSERT(pdat_fac);
+        depth = pdat_fac->getDefaultDepth();
+        ghost_width = pdat_fac->getGhostCellWidth().max();
+        TBOX_ASSERT(ghost_width == pdat_fac->getGhostCellWidth().min());
+        convertable = true;
```

Comment:
I think that we could refactor this internal block (which is common to every case in the `if/else if` chain) to something like
```cpp
template <typename FactoryType>
std::pair<int, int> // depth, ghost_width
get_characteristics(const PatchDescriptor<NDIM> &patch_descriptor,
                    const int idx)
{
    auto Pointer<FactoryType> = patch_descriptor->getPatchDataFactory(idx);
    depth = pdat_fac->getDefaultDepth();
    ghost_width = pdat_fac->getGhostCellWidth().max();
    TBOX_ASSERT(ghost_width == pdat_fac->getGhostCellWidth().min());
    return std::make_pair(depth, ghost_width);
}
```